### PR TITLE
Andrew/cuda mnist

### PR DESCRIPTION
**Pull request by andrewCodeDev**
Goal of this PR: make mnist CUDA compatible. Meanwhile:

- reorganizing device project structure into dedicated files
- make loss functions device compatible
- add more cuda/host kernels

### CODE REVIEW

**Comment by andrewCodeDev** [comment]
Build tests will fail - I need to finish up the `loss.zig` rewrite and make sure cross entropy is available on cuda.

File: examples/mnist/Makefile
Comment (andrewCodeDev) [inline examples/mnist/Makefile:1]: NTS - need to revert this.

File: src/cuda/CMakeLists.txt
Comment (andrewCodeDev) [inline src/cuda/CMakeLists.txt:1]: NTS - revert this before merge. This adds NVCC debug information.
